allennlp/training/single_task_trainer.py

```diff
@@ -0,0 +1,701 @@
+        else:
+            scheduler = None
+
+        num_serialized_models_to_keep = trainer_params.pop_int("num_serialized_models_to_keep", 20)
+        keep_serialized_model_every_num_seconds = trainer_params.pop_int(
+                "keep_serialized_model_every_num_seconds", None)
+        model_save_interval = trainer_params.pop_float("model_save_interval", None)
+        summary_interval = trainer_params.pop_int("summary_interval", 100)
+        histogram_interval = trainer_params.pop_int("histogram_interval", None)
+        should_log_parameter_statistics = trainer_params.pop_bool("should_log_parameter_statistics", True)
+        should_log_learning_rate = trainer_params.pop_bool("should_log_learning_rate", False)
+        log_batch_size_period = trainer_params.pop_int("log_batch_size_period", None)
+
+        evaluate_on_test = params.pop("evaluate_on_test", False)
+
+        if 'test' in all_datasets and evaluate_on_test:
+            evaluation_dataset = all_datasets['test']
+        elif 'test' in all_datasets:
+            logger.info("To evaluate on the test set after training, pass the "
+                        "'evaluate_on_test' flag, or use the 'allennlp evaluate' command.")
+            evaluation_dataset = None
+        else:
+            evaluation_dataset = None
+
+        params.assert_empty(cls.__name__)
+        return cls(model, optimizer, iterator,
+                   train_data, validation_data,
+                   patience=patience,
+                   validation_metric=validation_metric,
+                   validation_iterator=validation_iterator,
+                   shuffle=shuffle,
+                   num_epochs=num_epochs,
+                   serialization_dir=serialization_dir,
+                   cuda_device=cuda_device,
+                   grad_norm=grad_norm,
+                   grad_clipping=grad_clipping,
+                   learning_rate_scheduler=scheduler,
+                   num_serialized_models_to_keep=num_serialized_models_to_keep,
+                   keep_serialized_model_every_num_seconds=keep_serialized_model_every_num_seconds,
+                   model_save_interval=model_save_interval,
+                   summary_interval=summary_interval,
+                   histogram_interval=histogram_interval,
+                   should_log_parameter_statistics=should_log_parameter_statistics,
+                   should_log_learning_rate=should_log_learning_rate,
+                   log_batch_size_period=log_batch_size_period,
+                   evaluation_dataset=evaluation_dataset)
+
+    # Requires custom from_params.
+    @classmethod
+    def from_params(cls,  # type: ignore
+                    params: Params,
+                    serialization_dir: str,
+                    recover: bool = False) -> 'SingleTaskTrainer':
+        # pylint: disable=arguments-differ
+        all_datasets = training_util.datasets_from_params(params)
```

Comment:
where would you put it, then? 

I guess I could move it back to `commands.train` and put it behind a flag?

Reply:
Can't it just be reverted? Is there an issue with subclassed trainers having to respect this from_params interface if you want to use our `train` commands, where if you do, you just have to write your own command?

Reply:
I mean, in a lot of ways that defeats the point of this refactor, because you're basically saying that you can only use `allennlp train` if you have a training regime that looks just like the current one

Reply:
Or, perhaps another abstract class method on the base `Trainer` that has this logic?  I'm not certain how that would look, but there might be a way to make it work.

Reply:
Hmm, looking at this again, that's basically what `for_model` tries to do, isn't it?  Is there a way we can make that work without taking a `Params` object, so `from_params` happens first, or something?

Reply:
I don't think still being able to use our commands for different types of Trainers should be a goal of this refactoring at all. I don't think it is feasible to have a single command which will handle these types of variations. The point was to make it possible to have these Trainer variations without having to copy large chunks of code, which the refactoring definitely achieves. Whether you can use new Trainer subclasses with `allennlp train` is not relevant, in my opinion.

Reply:
Yeah, I can see that argument.  You're basically saying that the single-task training logic is hard-coded in the `train` command, and that's fine.  Logic for setting up other kinds of training should get its own command (either written by us or by someone else).  I would be fine with that outcome.

Reply:
I have made a bunch of changes to make things fully backward compatible (`Trainer` is still the supervised trainer and has the exact same public API as before). I do feel pretty strongly that we should allow people to `allennlp train` with alternative trainers. 

For example, once I get this PR merged, my plan is to write a real multitask trainer that hopefully the people on S2 who are using various hacks can use. I really don't want to tell them "by the way, you can't use `allennlp train` to train these models", because I think that makes for a bad (and surprising) user experience.

Reply:
I don't have a super strong opinion here either way.  But is it really that bad of an experience to tell them "use `allennlp train-multi-task` for the multi-task trainer"?  Just trying to clarify people's opinions and arguments.

Reply:
I think it's a much better experience to say "use `allennlp train` to train your model, vs "use `allennlp train` to train a vanilla supervised model, or `allennlp multi-task-train` to train a multi-task model, or some other command for one of our other trainers, or if you've written your own trainer subclass then you just don't get to run it using a known `allennlp` command."

I don't see what we gain by doing that, other than saving some `if / then` logic in commands.train to preserve backward compatibility.

Reply:
I wasn't advocating adding a new command - more that if you want to use a  trainer which doesn't adhere to the `allennlp train` api, such as requiring different constructor arguments etc, you _must_ use allennlp as a library. If other things need to change to make that easier, then we should fix things to make it easier.